### PR TITLE
feat(cli, sdk, svm-sdk): cli offchain quote commands

### PR DESCRIPTION
**Pull request by xeno097**
### Description

  Adds two new CLI commands for managing offchain-signed warp fee quotes against deployed `OffchainQuotedLinearFee` leaves on EVM and SVM:

  - **`hyperlane warp quote create`** — submits a standing signed quote against a deployed warp route. Takes `--chain`, `--destination`, `--recipient` (native format or `wildcard`), `--amount` (or `wildcard`), `--max-fee`, `--half-amount`,
   `--ttl` (seconds, must be > 0), and `--quote-signer-key`. CC routing-fee target router is auto-resolved from the warp config (specific-match preferred, falls back to `DEFAULT_ROUTER_KEY`).
  - **`hyperlane warp quote read`** — enumerates standing quotes on every supported chain in a warp route (or a single `--chain`). Optional `--recipients` array probes additional non-router recipients on protocols with non-enumerable
  storage (EVM); SVM ignores it (its BTreeMap enumerates on-chain). Output renders bytes32 sentinels as labels (`TARGET_ROUTER_NONE`, `DEFAULT_CROSS_COLLATERAL_ROUTER`, `WILDCARD_RECIPIENT`), with ISO timestamps and an `expired` flag.

  Underneath:
  - `@hyperlane-xyz/provider-sdk` — cross-VM `IRawWarpQuoteArtifactManager` / `IRawWarpQuoteReader` / `IRawWarpQuoteWriter` / `RawQuoteSigner` interfaces, `enumerateWarpQuoteCandidates`, `ReadStandingQuotesOpts`, sentinel constants
  - `@hyperlane-xyz/sdk` — `EvmQuoteArtifactManager` / `EvmQuoteWriter` / `EvmQuoteReader` / `EvmPrivateKeyQuoteSigner` against the contract's EIP-712 typed-data layout, plus a `buildFeeReadContextFromWarpDeployConfig` helper that bypasses
   AltVM token-type validation so EVM warps with `xerc20` / `fastCollateral` / etc. work
  - `@hyperlane-xyz/sealevel-sdk` — equivalent `SvmQuote*` surface against the SVM fee-program's `SubmitQuote` instruction; exports `resolveFeeSalt`
  - CLI bridges EVM↔AltVM in a single `factories.ts` switch (EVM doesn't implement `ProtocolProvider`), shared by both commands

  ### Drive-by changes

  None.

  ### Related issues

  N/A

  ### Backward compatibility

  Yes. All changes are additive — two new CLI commands and a new optional `ReadStandingQuotesOpts` param on the cross-VM reader interface (backwards-compatible). No infra tooling implications.

  ### Testing

  Unit tests + e2e tests:

  - **SDK unit/hardhat**: EVM reader (5 cases incl. `--recipients` discovery), EVM writer (7 cases incl. stale, expired, submitter-binding, transient-no-standing-storage)
  - **SVM SDK e2e**: writer (6 cases incl. stale, expired, transient-no-standing-PDA), reader (3 cases)
  - **provider-sdk unit**: `enumerateWarpQuoteCandidates` (8 cases incl. DEFAULT_KEY filter)
  - **CLI unit**: `resolveTargetRouterForVariant` CCRF decision tree (5 cases), `groupEntriesByScope` `expired` flag, `createQuoteArtifactManagerForChain` + `createDefaultQuoteSignerForChain` over all supported protocols
  - **CLI e2e**: EVM warp-quote (6 cases — plain leaf, RoutingFee, CCRF, `--chain` filter on source + destination, `--recipients` discovery, `--ttl=0` rejection), SVM warp-quote (4 cases — analogous)

  All green.

### CODE REVIEW

File: typescript/provider-sdk/src/quote.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    }
+    candidates.push({
+      destination,
+      recipient: WILDCARD_BYTES32,
+      targetRouter: WARP_TARGET_ROUTER_NONE,
+      amount: WARP_QUOTE_AMOUNT_WILDCARD,
+    });
+  }
+
+  for (const router of allRouters) {
```
Comment (paulbalaji) [inline typescript/provider-sdk/src/quote.ts:213]: The wildcard-destination loop emits candidates only with `targetRouter: WARP_TARGET_ROUTER_NONE` (and skips `DEFAULT_CROSS_COLLATERAL_FEE_ROUTER_KEY`), unlike the per-domain loop above which also emits the CC `(recipient: WILDCARD_BYTES32, targetRouter: router)` variant. SVM CC standing quotes registered under a wildcard *domain* with a concrete `targetRouter` would therefore never be enumerated, so `warp quote read` could silently miss them. Is the asymmetry intentional? If wildcard-domain CC quotes are possible on SVM, this loop should mirror the per-domain `targetRouter` emission.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        return key === destinationChainName;
+      });
+      if (routerEntry) {
+        const destRouterBytes32 = addressToBytes32(
+          routerEntry[1].address,
+          destinationProtocol,
+        );
+        if (destRoutes[destRouterBytes32]) return destRouterBytes32;
+      }
+      if (destRoutes[DEFAULT_CROSS_COLLATERAL_FEE_ROUTER_KEY]) {
```
Comment (paulbalaji) [inline typescript/cli/src/deploy/warp-quote.ts:324]: This CCRF fallback keys off `DEFAULT_CROSS_COLLATERAL_FEE_ROUTER_KEY` (provider-sdk), but `offchainQuotedLeaf.ts:61` (`resolveOffchainQuotedLeafAddress`) keys the *same* fallback off `DEFAULT_ROUTER_KEY` (sdk). They're two independent literal definitions of the identical bytes32 today, so it works — but target-router resolution and leaf lookup must agree, and editing one constant would silently break the create flow. Use a single canonical constant across `deploy/warp-quote.ts`, `read/warp-quote.ts`, and `offchainQuotedLeaf.ts`.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  // generated internally and never returned to the caller.
+  assert(
+    ttl > 0,
+    `--ttl must be > 0 (transient quotes are unusable from standalone create), got ${ttl}`,
+  );
+  const expiry = issuedAt + ttl;
+
+  logBlue(
+    `Submitting ${
+      ttl === 0 ? 'transient' : 'standing'
```
Comment (paulbalaji) [inline typescript/cli/src/deploy/warp-quote.ts]: Dead branch: `assert(ttl > 0, …)` above guarantees `ttl !== 0`, so `ttl === 0 ? 'transient' : 'standing'` always evaluates to `'standing'`. Replace with the literal `'standing'`.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+      // SVM submit_quote does NO cascade — the signed targetRouter must
+      // match an actually-configured route key on the fee account.
+      // EVM cascades at quote-read time, so it's tolerant of either
+      // specific or DEFAULT here. Prefer a specific match, fall back to
+      // DEFAULT_ROUTER_KEY.
+      const destRoutes = tokenFee.feeContracts[destinationChainName] ?? {};
+      const routers = localConfig.remoteRouters ?? {};
+      const routerEntry = Object.entries(routers).find(([key]) => {
+        const domain = multiProvider.tryGetDomainId(key);
+        if (domain !== null && domain !== undefined)
```
Comment (paulbalaji) [inline typescript/cli/src/deploy/warp-quote.ts]: `multiProvider.tryGetDomainId` returns `number | null`, never `undefined`, so the `&& domain !== undefined` half is redundant — `if (domain !== null)` suffices.

**Review by paulbalaji — COMMENTED** [review]
## Review Summary

Reviewed at `04bec1c18a` (stacked on `xeno/mailbox-upgrade-logic`). No prior reviews, comments, or inline threads existed at this commit.

Overall this is a well-structured, well-tested addition. EIP-712 typed-data is byte-correct against `AbstractOffchainQuoter.sol`, SVM is at parity with EVM, the type-cast audit is clean (only the canonical `as Record<string, unknown>` type-guard pattern appears), and the changeset is present and correct. The inline notes below are mostly minor/nit polish, with a few worth a closer look (CCRF default-router constant divergence, the wildcard-destination CC enumeration asymmetry, and raw 20-byte router addresses flowing into a bytes32 ABI slot).

### ⚠️ E2E tests are not wired into CI

The PR adds e2e tests and the description says "All green," but **none of the new e2e tests actually run in CI** — `grep warp-quote .github/workflows` returns nothing. Both runners select tests by an explicit env var in CI (the `**/*.e2e-test.ts` glob only applies to local runs):

- `typescript/cli/src/tests/ethereum/run-e2e-test.sh` / `sealevel/run-e2e-test.sh` run only `**/${CLI_E2E_TEST}.e2e-test.ts`, and CI always sets `CLI_E2E_TEST: ${{ matrix.test }}`. There is **no `warp-quote` entry** in either the `cli-evm-e2e-matrix` or `cli-svm-e2e-matrix` in `.github/workflows/test-cli-e2e.yml`.
- `typescript/svm-sdk/scripts/run-e2e-test.sh` *requires* `SVM_SDK_E2E_TEST` (errors if unset), and `warp-quote-reader` / `warp-quote-writer` are **not** in the `svm-sdk-e2e-matrix` in `.github/workflows/test-sdk-e2e.yml`.

So the green checks do not prove these tests pass. Suggest adding `warp-quote` to the EVM + SVM CLI e2e matrices and `warp-quote-reader`/`warp-quote-writer` to the SVM SDK e2e matrix. (Flagged inline on the svm-sdk runner; the workflow files themselves are outside this PR's diff — see below.)

## Observations Outside This PR (not in diff, can't inline)

- `.github/workflows/test-cli-e2e.yml` `cli-evm-e2e-matrix` (~line 75) and `cli-svm-e2e-matrix` (~line 542): add a `warp-quote` matrix entry to each so the new EVM/SVM CLI e2e tests run in CI.
- `.github/workflows/test-sdk-e2e.yml` `svm-sdk-e2e-matrix` (~line 165): add `warp-quote-reader` and `warp-quote-writer` so the new SVM SDK e2e tests run in CI.

## Nits (not individually inlined)

- `chainResolver.ts`: the `warp alt create` signs-on-SVM-only comment now sits above `resolveWarpQuoteCreateChains` after the insertion, but describes `resolveWarpAltCreateChains` below it — move it back down.
- `deploy/warp-quote.ts` `resolveTargetRouterForVariant`: consider the `default: { const _exhaustive: never = … }` guard its `factories.ts` siblings use.
- Bare `'wildcard'` input sentinel is repeated across `commands/warp-quote.ts` + `deploy/warp-quote.ts`; a shared `const` would keep parse logic and help text in sync.

File: typescript/cli/src/read/warp-quote.ts [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    const destKey =
+      multiProvider.tryGetChainName(entry.scope.destination) ??
+      String(entry.scope.destination);
+    const byRouter = (grouped[destKey] ??= {});
+    const byRecipient = (byRouter[
+      targetRouterLabel(entry.scope.targetRouter)
+    ] ??= {});
+    byRecipient[recipientLabel(entry.scope.recipient)] = {
+      amount:
+        entry.scope.amount.kind === 'wildcard'
```
Comment (paulbalaji) [inline typescript/cli/src/read/warp-quote.ts]: Raw string literal `'wildcard'` — prefer the enum `WarpQuoteAmountKind.wildcard` (already imported/used in `deploy/warp-quote.ts` and defined in provider-sdk) for refactor safety, per the codebase's enum-over-literal convention.

File: typescript/cli/src/read/warp-quote.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+          ? 'wildcard'
+          : entry.scope.amount.value.toString(),
+      maxFee: entry.params.maxFee.toString(),
+      halfAmount: entry.params.halfAmount.toString(),
+      issuedAt: unixSecToIsoString(entry.issuedAt),
+      expiry: unixSecToIsoString(entry.expiry),
+      expired: nowSec > entry.expiry,
```
Comment (paulbalaji) [inline typescript/cli/src/read/warp-quote.ts:260]: `nowSec > entry.expiry` uses strict `>`, so a quote at the exact boundary (`nowSec === entry.expiry`) renders as not-expired. Confirm `>` vs `>=` matches the on-chain expiry semantics, and add a boundary case to `groupEntriesByScope.test.ts` (currently only past/future are covered).

File: typescript/sdk/src/deploy/warp.ts [hunk 7]
```diff
@@ -85,6 +88,58 @@ const SUPPORTED_ALTVM_TOKEN_TYPES = new Set<TokenType>([
...
+
+  for (const [chainNameOrId, router] of Object.entries(
+    config.remoteRouters ?? {},
+  )) {
+    const domain = chainLookup.getDomainId(chainNameOrId);
+    if (isNullish(domain)) continue;
+    const existing = knownRoutersPerDomain[domain] ?? new Set();
+    knownRoutersPerDomain[domain] = new Set([
+      ...existing,
+      router.address,
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/warp.ts]: `router.address` (and `...routers` at line 134) are copied verbatim into `knownRoutersPerDomain`. `crossCollateralRouters` is typed `z.array(ZHash)` and `remoteRouters[].address` is also `ZHash`, both of which accept 20-byte EVM addresses. These values become candidate `recipient`/`targetRouter` entries that `EvmQuoteReader` passes straight into `contract.quotes(destination, recipient)` (a `bytes32` ABI slot) with no try/catch — a 20-byte value would make ethers throw `INVALID_ARGUMENT` and fail the entire read. Normalize to bytes32 (`addressToBytes32`) here, or guard/normalize in the reader.

File: typescript/cli/src/commands/warp-quote.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  },
+  handler: async ({ context, warpRouteId, chain, recipients, out }) => {
+    logCommandHeader('Hyperlane Warp Quote Read');
+    // `addressToBytes32` auto-detects protocol from the address format
+    // (EVM 0x-hex, Sealevel base58, etc.), so the CLI can convert here
+    // without knowing the per-destination protocols of the warp route.
+    const extraRecipients = new Set<string>();
+    for (const native of recipients ?? []) {
+      try {
+        extraRecipients.add(addressToBytes32(native));
```
Comment (paulbalaji) [inline typescript/cli/src/commands/warp-quote.ts:151]: Invalid `--recipients` entries are silently swallowed by the `catch {}`. A typo'd recipient is dropped with no warning, producing a false-empty read on EVM (where this is the only way to probe non-enumerable storage). At minimum log a warning naming the dropped input, per the "never swallow errors silently" guideline.

File: typescript/cli/src/quote/offchainQuotedLeaf.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+export function resolveOffchainQuotedLeafAddress(args: {
+  tokenFee: DerivedTokenFeeConfig;
+  destChainName: string;
+  targetRouterBytes32: string;
+}): string {
+  const { tokenFee, destChainName, targetRouterBytes32 } = args;
+  switch (tokenFee.type) {
+    case TokenFeeType.OffchainQuotedLinearFee:
```
Comment (paulbalaji) [inline typescript/cli/src/quote/offchainQuotedLeaf.ts:35]: For a plain `OffchainQuotedLinearFee` leaf, `create` returns the leaf address without checking that `--destination` is actually enrolled in the warp route. A valid-but-unenrolled destination (e.g. a typo that resolves to a real domain) would submit a standing quote keyed to a destination that normal `warp quote read` never enumerates — an effectively invisible, unusable quote. Consider validating the destination against the route's enrolled domains before submitting.

File: typescript/cli/src/context/strategies/chain/chainResolver.ts [hunk 10]
```diff
@@ -54,9 +54,12 @@ export async function resolveChains(
     case CommandType.WARP_READ:
     case CommandType.WARP_ALT_CHECK:
     case CommandType.WARP_ALT_READ:
+    case CommandType.WARP_QUOTE_READ:
```
Comment (paulbalaji) [inline typescript/cli/src/context/strategies/chain/chainResolver.ts]: `WARP_QUOTE_READ` routes to the shared `resolveWarpReadChains`, where the `argv.warpRouteId` branch overwrites `context.chains` with *all* route chains even when `--chain` was given. Since `warp quote read` always has a `warpRouteId`, a single-chain read still provisions providers/signers for every chain in the route — on a mixed EVM/SVM route, `warp quote read --chain <evm>` would still try to load AltVM providers and could fail/hang. (Behavior is inherited from the existing resolver, but the read command's own `--chain` filter implies single-chain intent that this undermines for provisioning.)

File: typescript/svm-sdk/scripts/run-e2e-test.sh [hunk 11]
```diff
@@ -3,7 +3,7 @@ set -e
 
 if [ -z "${SVM_SDK_E2E_TEST}" ]; then
   echo "Error: SVM_SDK_E2E_TEST env var is required."
-  echo "Available tests: ism, hook, mailbox, mailbox-program-upgrade, validator-announce, native-token, synthetic-token, collateral-token, cross-collateral-token, provider, read-token, program-upgrade, warp-alt-simulation-parity, warp-fee-config, warp-transfer-remote-with-fees"
+  echo "Available tests: ism, hook, mailbox, mailbox-program-upgrade, validator-announce, native-token, synthetic-token, collateral-token, cross-collateral-token, provider, read-token, program-upgrade, warp-alt-simulation-parity, warp-fee-config, warp-quote-reader, warp-quote-writer, warp-transfer-remote-with-fees"
```
Comment (paulbalaji) [inline typescript/svm-sdk/scripts/run-e2e-test.sh]: `warp-quote-reader` / `warp-quote-writer` are listed here and run when `SVM_SDK_E2E_TEST` is set, but they are not in the `svm-sdk-e2e-matrix` in `.github/workflows/test-sdk-e2e.yml`, so they never run in CI. Add both to that matrix.

**Review by paulbalaji — COMMENTED** [review]
Consolidated review at `20fad11`. Scope: increment vs `xeno/mailbox-upgrade-logic`.

Carry-over blocker: neither new CLI e2e suite is in the `test-cli-e2e.yml` matrices (cli-evm / cli-svm), so the 592 lines of new e2e never run in CI; only `test-sdk-e2e.yml` was wired (`20fad11` covers the SVM SDK reader/writer suites only).

**Still open, no replies:** misplaced "warp alt create signs on SVM chains only" comment in `chainResolver.ts`; missing default-case exhaustiveness guard in `resolveTargetRouterForVariant`'s switch; bare `'wildcard'` sentinel duplicated across `commands/` and `deploy/`. On the expiry-boundary thread: strict `>` is pinned via comment plus boundary tests, but on-chain parity is asserted by comment only; a quick cross-check against the contract/program semantics would let that thread close.

Stack note: the wildcard-CC-enumeration fix on the base branch was read-side only, which widens the create/read asymmetry behind the inline CC finding; the create-side fix belongs in this PR.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+      return WARP_TARGET_ROUTER_NONE;
+
+    case TokenFeeType.CrossCollateralRoutingFee: {
+      // SVM submit_quote does NO cascade — the signed targetRouter must
+      // match an actually-configured route key on the fee account.
+      // EVM cascades at quote-read time, so it's tolerant of either
+      // specific or DEFAULT here. Prefer a specific match, fall back to
+      // DEFAULT_ROUTER_KEY.
+      const destRoutes = tokenFee.feeContracts[destinationChainName] ?? {};
+      const routers = localConfig.remoteRouters ?? {};
```
Comment (paulbalaji) [inline typescript/cli/src/deploy/warp-quote.ts]: **[medium] CC quote creation cannot target router-keyed leaves.** This resolution consults only `localConfig.remoteRouters` then falls back to DEFAULT, never `crossCollateralRouters`, and there is no `--target-router` flag. Derived CC configs key `feeContracts[dest]` by exactly DEFAULT + CC routers, so router-keyed OQLF leaves are unreachable from `create`:

- CC-router-only leaf, no DEFAULT: misleading "no leaf for destination" throw, even though `read` displays the leaf.
- Both present: the quote silently lands in the DEFAULT leaf/PDA; transfers via that CC router resolve the specific scope first (EVM per-router-before-DEFAULT, SVM scope isolation, no cascade), so the quote never applies while the CLI prints success.

Per-router slots are the production pattern (CROSS/moonpay: DEFAULT/USDC/USDT/ctUSD/XO). The read path unions CC routers, so this is a create/read asymmetry introduced here. Fix: consult `crossCollateralRouters` in resolution and add a `--target-router` override. Unit tests currently cover only remoteRouter/DEFAULT keys.

File: typescript/cli/src/commands/warp-quote.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+      description: 'Linear fee curve `halfAmount` parameter (wei/lamports)',
+      demandOption: true,
+    },
+    ttl: {
+      type: 'number',
+      description:
+        'Time-to-live in seconds for the standing quote (expiry = now + ttl). Must be > 0 — transient quotes (ttl=0) are not usable from this standalone command because their on-chain storage is scoped to the create tx (EIP-1153 on EVM, payer-scoped PDA on SVM).',
+      demandOption: true,
+    },
+    'quote-signer-key': {
```
Comment (paulbalaji) [inline typescript/cli/src/commands/warp-quote.ts:80]: **[low]** `--quote-signer-key` is demandOption with no env fallback, so the quote-authority key must be passed on argv: shell history, `ps`, CI logs (the e2e helper interpolates it into a command line). The CLI's own `keyCommandOption` pattern defaults from env; a `HYP_QUOTE_SIGNER_KEY` default here is a one-liner.

File: typescript/sdk/src/quote/EvmQuoteWriter.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    const submitter = await this.txSigner.getAddress();
+    const salt = ethersUtils.hexlify(ethersUtils.randomBytes(32));
+
+    const sq = buildEvmSignedQuoteTuple(req, salt, submitter);
+    const signable = buildEvmSignedQuoteSignable(sq, chainId, this.feeAddress);
+    const { signature } = await this.quoteSigner.sign(signable);
+    const signatureHex = ethersUtils.hexlify(signature);
+
+    const tx = await contract.submitQuote(sq, signatureHex);
+    const receipt = await tx.wait();
```
Comment (paulbalaji) [inline typescript/sdk/src/quote/EvmQuoteWriter.ts:61]: **[low]** An equal-`issuedAt` resubmission is an on-chain no-op (`_storeStanding` returns false, no QuoteSubmitted event; SVM submit returns Ok) but this only awaits `tx.wait()` and the CLI prints "Quote submitted": corrected params are silently ignored. Narrow for interactive CLI (same-second re-run) but real for programmatic callers. Cheap fix: check for QuoteSubmitted in the receipt and warn on no-op.

File: typescript/provider-sdk/src/quote.ts (hunk 1 above)
Comment (xeno097) [reply typescript/provider-sdk/src/quote.ts:213]: Good catch — the wildcard-destination loop now mirrors the per-domain emission, also yielding the CC `(recipient: WILDCARD, targetRouter: router)` variant so wildcard-domain CC standing quotes are enumerated. Fixed in 88f336de90.

File: typescript/cli/src/deploy/warp-quote.ts (hunk 2 above)
Comment (xeno097) [reply typescript/cli/src/deploy/warp-quote.ts:324]: These are not independent literals: `DEFAULT_ROUTER_KEY` (sdk) is an alias re-export of `DEFAULT_CROSS_COLLATERAL_FEE_ROUTER_KEY` (provider-sdk) — one canonical definition. Leaf lookup and target-router resolution reference the same value, so editing it cannot desync them.

File: typescript/cli/src/deploy/warp-quote.ts (hunk 3 above)
Comment (xeno097) [reply typescript/cli/src/deploy/warp-quote.ts]: Fixed in 67b5c0ab98 — replaced the dead ternary with the literal `'standing'` (the `assert(ttl > 0)` above already guarantees it).

File: typescript/cli/src/deploy/warp-quote.ts (hunk 4 above)
Comment (xeno097) [reply typescript/cli/src/deploy/warp-quote.ts]: Fixed in 94b5a08221 — dropped the redundant `undefined` half; the check is now just `!isNullish(domain)`.

File: typescript/cli/src/read/warp-quote.ts (hunk 5 above)
Comment (xeno097) [reply typescript/cli/src/read/warp-quote.ts]: Fixed in 1d47ec8ce9 — now uses `WarpQuoteAmountKind.wildcard` instead of the raw literal.

File: typescript/cli/src/read/warp-quote.ts (hunk 6 above)
Comment (xeno097) [reply typescript/cli/src/read/warp-quote.ts:260]: Confirmed against the contract: `AbstractOffchainQuoter.submitQuote` reverts only when `uint48(block.timestamp) > sq.expiry` (AbstractOffchainQuoter.sol:89), so a quote at exactly `now === expiry` is still valid on-chain — the strict `>` matches. Boundary case added in 5a186c9680.

File: typescript/sdk/src/deploy/warp.ts (hunk 7 above)
Comment (xeno097) [reply typescript/sdk/src/deploy/warp.ts]: Fixed in 37f2ca9044 — router addresses are normalized via `addressToBytes32` before entering `knownRoutersPerDomain`, so 20-byte values no longer reach the `bytes32` ABI slot.

File: typescript/cli/src/commands/warp-quote.ts (hunk 8 above)
Comment (xeno097) [reply typescript/cli/src/commands/warp-quote.ts:151]: Fixed in 7e1c91024c — invalid `--recipients` entries now emit a warning naming the dropped input instead of being silently swallowed.

File: typescript/cli/src/quote/offchainQuotedLeaf.ts (hunk 9 above)
Comment (xeno097) [reply typescript/cli/src/quote/offchainQuotedLeaf.ts:35]: Fixed in dca98dd26a — `create` now validates the destination is enrolled in the route before submitting, so a quote can no longer be keyed to an unenumerable destination.

File: typescript/cli/src/context/strategies/chain/chainResolver.ts (hunk 10 above)
Comment (xeno097) [reply typescript/cli/src/context/strategies/chain/chainResolver.ts]: Fixed in 859130e0c0 — `warp quote read` now scopes chain resolution to `--chain`, so a single-chain read no longer provisions providers/signers for every chain in the route.

File: typescript/svm-sdk/scripts/run-e2e-test.sh (hunk 11 above)
Comment (xeno097) [reply typescript/svm-sdk/scripts/run-e2e-test.sh]: Fixed — `warp-quote-reader`/`warp-quote-writer` added to the `svm-sdk-e2e-matrix` (971d9b4d89), and the EVM + SVM CLI e2e suites wired into `cli-evm-e2e-matrix` and `cli-svm-e2e-matrix` (49e0f53f77).

File: typescript/cli/src/deploy/warp-quote.ts (hunk 12 above)
Comment (xeno097) [reply typescript/cli/src/deploy/warp-quote.ts]: Fixed in cae46a58da — CCRF resolution now also consults the destination's `crossCollateralRouters`, and a new `--target-router` flag (destination-native address, validated against the configured leaves) targets a specific router-keyed leaf. Ambiguous auto-resolution (multiple matching leaves, no DEFAULT) throws and asks for `--target-router`. Unit tests cover CC-router match, explicit override, missing-leaf, and ambiguity.

File: typescript/cli/src/commands/warp-quote.ts (hunk 13 above)
Comment (xeno097) [reply typescript/cli/src/commands/warp-quote.ts:80]: Fixed in 90633ded9a — `--quote-signer-key` now defaults from `HYP_QUOTE_SIGNER_KEY` (matching the `keyCommandOption` pattern), so the key no longer has to be passed on argv.

File: typescript/sdk/src/quote/EvmQuoteWriter.ts (hunk 14 above)
Comment (xeno097) [reply typescript/sdk/src/quote/EvmQuoteWriter.ts:61]: Fixed in ca2315dac9 — `EvmQuoteWriter` now inspects the receipt for `QuoteSubmitted` and surfaces `standingStored` on `SubmittedWarpQuote`; the CLI warns on a no-op instead of reporting success. Covered by a hardhat test in 52baa9d5d4.

**Review by paulbalaji — COMMENTED** [review]
Rereviewed at `52baa9d5d4`. All prior findings are addressed: EVM/SVM CLI quote e2es and SVM reader/writer e2es are wired into their matrices; wildcard-domain CC candidates are enumerated; expiry uses the contract's strict `>` boundary with tests; invalid recipients warn; unenrolled destinations fail fast; the signer key has an env fallback; and equal-issuedAt EVM no-ops are surfaced. The target-router follow-up also handles cross-collateral router-keyed leaves and ambiguous matches cleanly.

No new blockers. Focused validation passed: 112 CLI tests, 89 provider-sdk tests, and the relevant SDK/deploy tests on this exact head. Because the branch is draft, GitHub has only run peripheral checks at this head; the newly wired e2es still need one substantive exact-head CI run before undraft.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    destinationProtocol,
+    explicitTargetRouter,
+  } = args;
+  switch (tokenFee.type) {
+    case TokenFeeType.LinearFee:
+    case TokenFeeType.RegressiveFee:
+    case TokenFeeType.ProgressiveFee:
+    case TokenFeeType.OffchainQuotedLinearFee:
+    case TokenFeeType.RoutingFee:
+      return WARP_TARGET_ROUTER_NONE;
```
Comment (paulbalaji) [inline typescript/cli/src/deploy/warp-quote.ts:265]: `--target-router` is silently ignored for every non-CrossCollateral fee type: the `LinearFee | RegressiveFee | ProgressiveFee | OffchainQuotedLinearFee | RoutingFee` arm returns `WARP_TARGET_ROUTER_NONE` without consulting `explicitTargetRouter`, so a user who passes `--target-router` on e.g. a `RoutingFee` route gets no error and no effect — the quote quietly lands on the `TARGET_ROUTER_NONE` leaf. Per the fail-fast convention, consider asserting `explicitTargetRouter == null` in the non-CC arm (`--target-router only applies to CrossCollateralRoutingFee`).

Related minor point: `--amount` is `demandOption`, but since this command only writes standing quotes (`ttl > 0`) and both writers assert `amount.kind === wildcard` for standing quotes, a concrete `--amount` can never succeed — it always fails deep in `submitQuote`. Both low-severity ergonomics nits, not correctness bugs.

**Review by paulbalaji — COMMENTED** [review]
Rereviewed at `52baa9d5d4` (HEAD unchanged since my last pass). All 17 prior threads are resolved at HEAD: wildcard-domain CC enumeration (`recipient=router` so it survives the fully-wildcard filter), bytes32 router normalization, strict-`>` expiry documented against `AbstractOffchainQuoter.sol` + boundary test, invalid-recipient warning, unenrolled-destination fast-fail, `HYP_QUOTE_SIGNER_KEY` fallback, equal-issuedAt no-op surfacing via receipt check, CC target-router resolution with `--target-router` + ambiguity error + `never` exhaustiveness guard, `--chain` provider scoping, and the CLI/SVM-SDK e2e matrix wiring. `DEFAULT_ROUTER_KEY` aliasing checks out (single canonical constant, cannot desync).

Fresh checklist pass is otherwise clean — cast audit finds only canonical type-guard patterns, and EVM + SVM e2e cover create/read across plain/Routing/CC fee shapes. One new low-severity ergonomics nit inline; no blockers.

**Review by paulbalaji — APPROVED** [review]
Rereviewed at `52baa9d5d4` after reconciling current/outdated threads and author replies. Prior correctness findings are fixed at HEAD: wildcard CC enumeration, router normalization/resolution, expiry semantics, destination validation, signer-key handling, no-op reporting, chain scoping, and E2E matrix wiring. Focused provider-sdk validation passed (89 tests), and the prior exact-head CLI/SDK focused runs remain applicable because HEAD is unchanged. No blockers. The existing low-severity line-265 ergonomics thread remains: fail fast when `--target-router` is supplied to non-CC fees, and avoid accepting a concrete `--amount` for this standing-only command. Newly wired E2Es still need substantive exact-head CI before merge. Approving.

File: typescript/cli/src/deploy/warp-quote.ts [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+import { resolveOffchainQuotedLeafAddress } from '../quote/offchainQuotedLeaf.js';
+import { deriveWarpRouteConfigForChain } from '../read/warp.js';
+import { getWarpCoreConfigOrExit } from '../utils/warp.js';
+
+function parseBigIntFlag(name: string, raw: string): bigint {
+  try {
+    return BigInt(raw);
+  } catch {
+    throw new Error(`Invalid bigint value for --${name}: "${raw}"`);
+  }
```
[the remark this answers is not in the document]
Comment (xeno097) [reply typescript/cli/src/deploy/warp-quote.ts:41]: no negatives are allowed
